src/analyses/cfg_dominators.h

```diff
@@ -44,124 +238,271 @@ class cfg_dominators_templatet
 
 protected:
   void initialise(P &program);
-  void fixedpoint(P &program);
-};
-
-/// Print the result of the dominator computation
-template <class P, class T, bool post_dom>
-std::ostream &operator << (
-  std::ostream &out,
-  const cfg_dominators_templatet<P, T, post_dom> &cfg_dominators)
-{
-  cfg_dominators.output(out);
-  return out;
-}
-
-/// Compute dominators
-template <class P, class T, bool post_dom>
-void cfg_dominators_templatet<P, T, post_dom>::operator()(P &program)
-{
-  initialise(program);
-  fixedpoint(program);
-}
 
-/// Initialises the elements of the fixed point analysis
-template <class P, class T, bool post_dom>
-void cfg_dominators_templatet<P, T, post_dom>::initialise(P &program)
-{
-  cfg(program);
-}
+  struct fixedpointt
+  {
+    explicit fixedpointt(cfg_dominators_templatet &cfg_dominators)
+      : cfg_dominators(cfg_dominators),
+        dfsCounter(0),
+        parent(cfg_dominators.cfg.size() + 1),
+        ancestor(cfg_dominators.cfg.size() + 1),
+        child(cfg_dominators.cfg.size() + 1),
+        vertex(cfg_dominators.cfg.size() + 1),
+        dom(cfg_dominators.cfg.size() + 1),
+        label(cfg_dominators.cfg.size() + 1),
+        semi(cfg_dominators.cfg.size() + 1),
+        size(cfg_dominators.cfg.size() + 1),
+        bucket(cfg_dominators.cfg.size() + 1)
+    {
+    }
 
-/// Computes the MOP for the dominator analysis
-template <class P, class T, bool post_dom>
-void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
-{
-  std::list<T> worklist;
+    void fixedpoint(P &program);
 
-  if(cfg.nodes_empty(program))
-    return;
+  private:
+    cfg_dominators_templatet &cfg_dominators;
+    node_indext dfsCounter;
+    std::vector<node_indext> parent, ancestor, child, vertex, dom;
 
-  if(post_dom)
-    entry_node=cfg.get_last_node(program);
-  else
-    entry_node=cfg.get_first_node(program);
-  typename cfgt::nodet &n=cfg[cfg.entry_map[entry_node]];
-  n.dominators.insert(entry_node);
+    std::vector<node_indext> label, semi, size;
 
-  for(typename cfgt::edgest::const_iterator
-      s_it=(post_dom?n.in:n.out).begin();
-      s_it!=(post_dom?n.in:n.out).end();
-      ++s_it)
-    worklist.push_back(cfg[s_it->first].PC);
+    std::vector<std::set<node_indext>> bucket;
 
-  while(!worklist.empty())
-  {
-    // get node from worklist
-    T current=worklist.front();
-    worklist.pop_front();
+    T get_entry_node(P &program)
+    {
+      if(post_dom)
+      {
+        return cfg_dominators.cfg.get_last_node(program);
+      }
+      else
+      {
+        return cfg_dominators.cfg.get_first_node(program);
+      }
+    };
 
-    bool changed=false;
-    typename cfgt::nodet &node=cfg[cfg.entry_map[current]];
-    if(node.dominators.empty())
+    void dfs(node_indext root)
```

Comment:
Is this not depth-first search on a `grapht`? If it isn't, please add comments. If it is, please move the implementation to util/graph.h.

Reply:
It is dfs on a `grapht`, yes. How would you like it `moved` to graph.h? Like

`template<typename Graph, typename Function>
void depth_first_search(const Graph& graph, typename graph::index_t start_node, Function function);`?

Reply:
Couldn't it become a member of `grapht`?

Reply:
Sure, what I meant was if you wanted it to take a "per node work" function as a parameter.

Reply:
Yes, your suggestion is good, except I’d recommend making it a member function.